### ISSUE TITLE
使用paddleseg训练ocrnet，在验证的时候会占用极大的显存之后退出。

### ISSUE THREAD

**Issue by 1wang11lijian1** [PaddlePaddle/PaddleSeg#3495]
### 问题确认 Search before asking

- [X] 我已经搜索过问题，但是没有找到解答。I have searched the question and found no related answer.

### 请提出你的问题 Please ask your question

本人在笔记本4060显卡上使用paddleseg训练ocrnet，设置批次大小为1，训练过程没有问题，但是验证的时候会占用极大的显存之后退出。

接下来是训练的过程，在第一轮验证的时候会经过两分钟时间显存剧增之后意外退出
![屏幕截图 2023-09-11 155828](https://github.com/PaddlePaddle/PaddleSeg/assets/101158990/0f25a465-cb8c-4b14-956c-845acb9a2e38)

下面是训练过程占用的显存
![屏幕截图 2023-09-11 160300](https://github.com/PaddlePaddle/PaddleSeg/assets/101158990/8ef0cec2-5cea-4a1d-85de-68abe7ca6289)

下面是验证过程占用的显存

![屏幕截图 2023-09-11 160347](https://github.com/PaddlePaddle/PaddleSeg/assets/101158990/c6e623ae-e892-4e83-ac63-4ec63b621f4f)

希望有大佬能解决一下这个问题！

**Comment by shiyutang** [comment]
评估的时候默认使用原图大小，如果超出显存，可以尝试在eval的部分增加和训练一样的图片裁剪和缩放等增强。

**Comment by 1wang11lijian1** [comment]
@shiyutang 好的非常感谢你的回答，我明白了你的解释，既然这样我想请问一下，我在cityscapes.yml配置文件的验证部分添加验证图片裁剪和缩放手段，之后在val.py需要修改什么地方吗？因为我按照前面的修改遇到了图像数据中的img跟label尺寸不一致的问题，img的尺寸是裁剪缩放后的而label是原图像的尺寸。希望能够再次得到您的帮助。

具体cityscapes.yml配置文件如下

```
batch_size: 1
iters: 10000

train_dataset:
  type: Cityscapes
  dataset_root: E:/weizhun/WZDeep/Dataset/Keyboard_data_En_cityscapes
  transforms:
    - type: ResizeStepScaling
      min_scale_factor: 0.5
      max_scale_factor: 2.0
      scale_step_size: 0.25
    - type: RandomPaddingCrop
      crop_size: [512, 375]
    - type: RandomHorizontalFlip
    - type: RandomDistort
      brightness_range: 0.4
      contrast_range: 0.4
      saturation_range: 0.4
    - type: Normalize
  mode: train

val_dataset:
  type: Cityscapes
  dataset_root: E:/weizhun/WZDeep/Dataset/Keyboard_data_En_cityscapes
  transforms:
    - type: ResizeStepScaling
      min_scale_factor: 0.5
      max_scale_factor: 2.0
      scale_step_size: 0.25
    - type: RandomPaddingCrop
      crop_size: [512, 375]
    - type: RandomHorizontalFlip
    - type: RandomDistort
      brightness_range: 0.4
      contrast_range: 0.4
      saturation_range: 0.4
    - type: Normalize
  mode: val


optimizer:
  type: SGD
  momentum: 0.9
  weight_decay: 4.0e-5

lr_scheduler:
  type: PolynomialDecay
  learning_rate: 0.01
  end_lr: 0
  power: 0.9

loss:
  types:
    - type: CrossEntropyLoss
  coef: [1]
```

**Comment by shiyutang** [comment]
针对你的需要，需要在L157增加data['gt_fields'].append('label')，保证可以同步修改。
https://github.com/PaddlePaddle/PaddleSeg/blob/63f95e61bedd0612bf004c60de94ac1a3d2ee4c5/paddleseg/datasets/dataset.py#L156
![image](https://github.com/PaddlePaddle/PaddleSeg/assets/34859558/a1134eb8-f428-494a-8752-f9cdbb46fb09)

**Comment by 1wang11lijian1** [comment]
好的谢谢你的解答，这个方法非常有效。

**Comment by 1wang11lijian1** [comment]
@shiyutang 您好抱歉再打扰你一次，我在predict几张原图的时候会出现尺寸不一致的问题，然后我尝试将cityscapes.yml文件增加了‘predict_dataset:’字段负责专门预测图片，相比val_dataset字段只缺少图片裁剪和缩放等增强操作，更改代码如下。其实这个操作相当于用原图尺寸预测，但是没有任何预测结果。所以请问预测的话是否需要用原图或者是将原图进行增强等操作？如何将原图进行增强等操作？

```
predict_dataset:
  type: Cityscapes
  dataset_root: data/Keyboard_data_En_cityscapes
  transforms:
    - type: Normalize
  mode: predict
```

**Comment by shiyutang** [comment]
你自己加上的配置字段我们没有代码读取是不会有效果的。你说的尺寸不一致是什么问题呢？ 预测的时候就按照原来图片的尺度就可以了。

**Comment by 1wang11lijian1** [comment]
@shiyutang  你好，我这边的情况是这样的，使用的数据集是4096x3000大小，本来我训练的时候通过剪裁缩放到了512x375大小，训练的第一种情况就是您这边默认的验证方式图像尺寸为原图也就是4096x3000的，训练之后再单独predict测试图片也是没有问题的，但是这种情况训练以及predict中需要消耗的显存过多，大概要占22G的GPU内存，所以基于此就考虑了第二种情况就是更改默认验证阶段的图像尺寸为512x375，配置文件更改如下，在paddleseg/cvlibs/builder.py文件中我也跟着修改了相关内容，这样一来消耗显存就大大减少，但是在生成掩膜的过程中会出现尺寸不一致问题，问题出在paddleseg/utils/visualize.py 46行 `vis_result = cv2.addWeighted(im, weight, pseudo_img, 1 - weight, 0)`，因为我”val_dataset“的字段里图像尺寸为512x375，而这里面的`im`为原图尺寸也就是4096x3000，所以在这步计算中会出错。因此我想请问更改了验证过程的图像输入大小之后，验证以及预测过程中的如何在比如说512x375上能够正常预测？

`batch_size: 1
iters: 80000

train_dataset:
  type: Cityscapes
  dataset_root: data/Keyboard_data_En_cityscapes
  transforms:
    - type: ResizeStepScaling
      min_scale_factor: 0.5
      max_scale_factor: 2.0
      scale_step_size: 0.25
    - type: RandomPaddingCrop
      crop_size: [512, 375]
    - type: RandomHorizontalFlip
    - type: RandomDistort
      brightness_range: 0.4
      contrast_range: 0.4
      saturation_range: 0.4
    - type: Normalize
  mode: train

val_dataset:
  type: Cityscapes
  dataset_root: data/Keyboard_data_En_cityscapes
  transforms:
    - type: ResizeStepScaling
      min_scale_factor: 0.5
      max_scale_factor: 2.0
      scale_step_size: 0.25
    - type: RandomPaddingCrop
      crop_size: [ 512, 375 ]
    - type: RandomHorizontalFlip
    - type: RandomDistort
      brightness_range: 0.4
      contrast_range: 0.4
      saturation_range: 0.4
    - type: Normalize
  mode: val

predict_dataset:
  type: Cityscapes
  dataset_root: data/Keyboard_data_En_cityscapes
  transforms:
    - type: Normalize
  mode: predict

optimizer:
  type: SGD
  momentum: 0.9
  weight_decay: 4.0e-5

lr_scheduler:
  type: PolynomialDecay
  learning_rate: 0.01
  end_lr: 0
  power: 0.9

loss:
  types:
    - type: CrossEntropyLoss
  coef: [1]
`

**Comment by 1liuren** [comment]
我也这样设置了，但是在单卡训练和验证时都没有问题，就是多卡时，开始验证就会卡住不动[2024/10/28 15:43:47] INFO: Start evaluating (total_samples: 2183, total_iters: 1092)...

**Comment by TingquanGao** [comment]
Thanks for this issue. As it has been inactive for a long time, we would close it. If you has any questions, please feel free to reopen or new issue, and we will follow up and resolve it.
